File: src/omnia/data/structure/tuple/ImmutableTriplet.java
package omnia.data.structure.tuple;

import java.util.Iterator;
import java.util.OptionalInt;
import java.util.stream.Stream;
import omnia.data.structure.immutable.ImmutableList;

class ImmutableTriplet<T> extends ImmutableTriple<T, T, T> implements Triplet<T> {

  ImmutableTriplet(T first, T second, T third) {
    super(first, second, third);
  }

  @Override
  public Couplet<T> dropFirst() {
    return Tuplet.of(second(), third());
  }

  @Override
  public Couplet<T> dropSecond() {
    return Tuplet.of(first(), third());
  }

  @Override
  public Couplet<T> dropThird() {
    return Tuplet.of(first(), second());
  }

  @Override
  public Quadruplet<T> concat(T object) {
    return Tuplet.of(first(), second(), third(), object);
  }

  @Override
  public Quintuplet<T> concat(Couple<T, T> other) {
    return Tuplet.of(first(), second(), third(), other.first(), other.second());
  }

  @Override
  public Sextuplet<T> concat(Triple<T, T, T> other) {
    return Tuplet.of(first(), second(), third(), other.first(), other.second(), other.third());
  }

  @Override
  public Septuplet<T> concat(Quadruple<T, T, T, T> other) {
    return Tuplet.of(first(), second(), third(), other.first(), other.second(), other.third(), other.fourth());
  }

  @Override
  public Octuplet<T> concat(Quintuple<T, T, T, T, T> other) {
    return Tuplet.of(first(), second(), third(), other.first(), other.second(), other.third(), other.fourth(), other.fifth());
  }

  @Override
  public Nonuplet<T> concat(Sextuple<T, T, T, T, T, T> other) {
    return Tuplet.of(first(), second(), third(), other.first(), other.second(), other.third(), other.fourth(), other.fifth(), other.sixth());
  }

  @Override
  public Decuplet<T> concat(Septuple<T, T, T, T, T, T, T> other) {
    return Tuplet.of(first(), second(), third(), other.first(), other.second(), other.third(), other.fourth(), other.fifth(), other.sixth(), other.seventh());
  }

  @Override
  public Iterator<T> iterator() {
    return toActualList().iterator();
  }

  @Override
  public T itemAt(int index) {
    return toActualList().itemAt(index);
  }

  @Override
  public OptionalInt indexOf(Object item) {
    return toActualList().indexOf(item);
  }

  @Override
  public Stream<T> stream() {
    return toActualList().stream();
  }

  @Override
  public boolean containsUnknownTyped(Object element) {
    return toActualList().containsUnknownTyped(element);
  }

  private ImmutableList<T> toActualList() {
    return ImmutableList.of(first(), second(), third());
  }

}

File: src/omnia/data/structure/tuple/ImmutableTriplet.kt
package omnia.data.structure.tuple

import java.util.function.Function
import omnia.data.structure.immutable.ImmutableList

internal class ImmutableTriplet<T : Any>(first: T, second: T, third: T) :
    ImmutableTriple<T, T, T>(first, second, third), Triplet<T> {

  override fun dropFirst(): Couplet<T> {
    return Tuplet.of(second(), third())
  }

  override fun <R : Any> map(mapper: Function<in T, out R>): Triplet<R> {
    return Tuplet.of(mapper.apply(first()), mapper.apply(second()), mapper.apply(third()))
  }

  override fun dropSecond(): Couplet<T> {
    return Tuplet.of(first(), third())
  }

  override fun dropThird(): Couplet<T> {
    return Tuplet.of(first(), second())
  }

  override fun concat(`object`: T): Quadruplet<T> {
    return Tuplet.of(first(), second(), third(), `object`)
  }

  override fun concat(other: Couple<T, T>): Quintuplet<T> {
    return Tuplet.of(first(), second(), third(), other.first(), other.second())
  }

  override fun concat(other: Triple<T, T, T>): Sextuplet<T> {
    return Tuplet.of(first(), second(), third(), other.first(), other.second(), other.third())
  }

  override fun concat(other: Quadruple<T, T, T, T>): Septuplet<T> {
    return Tuplet.of(
        first(),
        second(),
        third(),
        other.first(),
        other.second(),
        other.third(),
        other.fourth()
    )
  }

  override fun concat(other: Quintuple<T, T, T, T, T>): Octuplet<T> {
    return Tuplet.of(
        first(),
        second(),
        third(),
        other.first(),
        other.second(),
        other.third(),
        other.fourth(),
        other.fifth()
    )
  }

  override fun concat(other: Sextuple<T, T, T, T, T, T>): Nonuplet<T> {
    return Tuplet.of(
        first(),
        second(),
        third(),
        other.first(),
        other.second(),
        other.third(),
        other.fourth(),
        other.fifth(),
        other.sixth()
    )
  }

  override fun concat(other: Septuple<T, T, T, T, T, T, T>): Decuplet<T> {
    return Tuplet.of(
        first(),
        second(),
        third(),
        other.first(),
        other.second(),
        other.third(),
        other.fourth(),
        other.fifth(),
        other.sixth(),
        other.seventh()
    )
  }

  override fun iterator(): Iterator<T> {
    return toActualList().iterator()
  }

  override fun itemAt(index: Int): T {
    return toActualList().itemAt(index)
  }

  override fun indexOf(item: Any?): Int? {
    return toActualList().indexOf(item)
  }

  override fun containsUnknownTyped(item: Any?): Boolean {
    return toActualList().containsUnknownTyped(item)
  }

  private fun toActualList(): ImmutableList<T> {
    return ImmutableList.of(first(), second(), third())
  }
}